File: ydsz-desktop/src/contracts/astGrep.ts
/**
 * @file AST-Grep 契约模块
 *
 * 定义 AST-Grep 结构搜索相关的数据契约，与后端
 * `YdszRpcRoutercontracts::ast_grep` 保持字段一致。
 *
 * ## 核心契约
 *
 * - `AstGrepPreset`：预设模式 ID（snake_case 字符串联合）
 * - `AstGrepPresetInfo`：预设模式详情（id / displayName / supportedLanguages）
 * - `AstGrepLanguage`：目标语言（typescript / javascript / rust / python）
 * - `AstGrepCapture`：S-expression 查询的捕获项（name / text）
 * - `AstGrepMatch`：单条匹配命中（file / line / column / text / nodeKind / captures）
 * - `AstGrepCompiledPattern`：模式编译结果（s-expression + captures）
 *
 * ## RPC 输入
 *
 * - `AstGrepFindByNodeKindInput`：`indexer.astGrepFindByNodeKind`
 * - `AstGrepFindByQueryInput`：`indexer.astGrepFindByQuery`
 * - `AstGrepFindByNameInput`：`indexer.astGrepFindByName`（mode: calls | references）
 * - `AstGrepListPresetsInput`：`indexer.astGrepListPresets`
 * - `AstGrepCompilePatternInput`：`indexer.astGrepCompilePattern`
 * - `AstGrepRewriteInput`：`indexer.astGrepRewrite`
 *
 * ## RPC 输出
 *
 * - `AstGrepRewriteResult`：替换结果（newContent / replacements / matchLocations）
 *
 * ## 双路径调用
 *
 * 同一组契约在两条路径上都可用：
 * - Tauri 命令路径（`useComposerAstGrepSearch` 直接 `invoke`）
 * - WebSocket 路径（`useComposerAstGrepSearch` 走 `readNativeApi().indexer.*`）
 *
 * 字段顺序、命名均与 Rust 端 specta::Type 导出一致；TS 端通过
 * `import type` 复用本文件生成的类型。
 */

import { Schema } from "effect";
import { TrimmedNonEmptyString } from "./baseSchemas";

// ===========================================================================
// 预设枚举（snake_case 字符串联合）
// ===========================================================================

/** AST-Grep 预设模式（与 Rust 端 `AstGrepPreset` 一一对应） */
export const AstGrepPreset = Schema.Literal(
  /** 所有 `console.log(...)` 调用 */
  "console_log",
  /** 所有 `console.error(...)` 调用 */
  "console_error",
  /** 所有 `await fetch(...)` 调用 */
  "await_fetch",
  /** 所有 `try { ... } catch { ... }` 语句 */
  "try_catch",
  /** 所有 `TODO(...)` / `FIXME(...)` 注释 */
  "todo_comment",
  /** Rust: 所有 `unwrap()` 调用 */
  "rust_unwrap",
  /** Rust: 所有 `expect("...")` 调用 */
  "rust_expect",
  /** Python: 所有 `print(...)` 调用 */
  "py_print",
  /** Python: 所有 `except:` 异常处理 */
  "py_except",
);
export type AstGrepPreset = typeof AstGrepPreset.Type;

/** AST-Grep 预设模式详情 */
export const AstGrepPresetInfo = Schema.Struct({
  /** preset id（snake_case） */
  id: AstGrepPreset,
  /** 人类可读名 */
  displayName: Schema.String,
  /** 支持的语言列表（"typescript" / "javascript" / "rust" / "python"） */
  supportedLanguages: Schema.Array(Schema.String),
});
export type AstGrepPresetInfo = typeof AstGrepPreset.Type;

/** AST-Grep 目标语言 */
export const AstGrepLanguage = Schema.Literal(
  "typescript",
  "javascript",
  "rust",
  "python",
);
export type AstGrepLanguage = typeof AstGrepLanguage.Type;

// ===========================================================================
// 匹配命中
// ===========================================================================

/** 捕获项：S-expression 查询中 `@name` 节点对应的文本 */
export const AstGrepCapture = Schema.Struct({
  /** 捕获名 */
  name: TrimmedNonEmptyString,
  /** 捕获文本 */
  text: Schema.String,
});
export type AstGrepCapture = typeof AstGrepCapture.Type;

/** 单条匹配命中 */
export const AstGrepMatch = Schema.Struct({
  /** 命中的源文件（绝对路径或相对工作区根） */
  file: TrimmedNonEmptyString,
  /** 起始行（1-based） */
  line: Schema.Number,
  /** 起始列（1-based） */
  column: Schema.Number,
  /** 起始字节偏移 */
  startByte: Schema.Number,
  /** 结束字节偏移 */
  endByte: Schema.Number,
  /** 命中的原始文本 */
  text: Schema.String,
  /** 命中的节点类型 */
  nodeKind: TrimmedNonEmptyString,
  /** 捕获名 → 文本（仅 S-expression 查询使用） */
  captures: Schema.Array(AstGrepCapture),
});
export type AstGrepMatch = typeof AstGrepMatch.Type;

// ===========================================================================
// 模式编译
// ===========================================================================

/** 模式编译结果 */
export const AstGrepCompiledPattern = Schema.Struct({
  /** tree-sitter S-expression */
  query: TrimmedNonEmptyString,
  /** 顶层 capture 名（按出现顺序） */
  captures: Schema.Array(TrimmedNonEmptyString),
});
export type AstGrepCompiledPattern = typeof AstGrepCompiledPattern.Type;

// ===========================================================================
// 检索模式
// ===========================================================================

/** `astGrepFindByName` 检索模式 */
export const AstGrepFindMode = Schema.Literal("calls", "references");
export type AstGrepFindMode = typeof AstGrepFindMode.Type;

// ===========================================================================
// RPC 输入
// ===========================================================================

/** `indexer.astGrepFindByNodeKind` 输入 */
export const AstGrepFindByNodeKindInput = Schema.Struct({
  /** 工作区根目录 */
  workspaceRoot: TrimmedNonEmptyString,
  /** 节点类型，如 `try_statement` / `call_expression` / `await_expression` */
  kind: TrimmedNonEmptyString,
});
export type AstGrepFindByNodeKindInput = typeof AstGrepFindByNodeKindInput.Type;

/** `indexer.astGrepFindByQuery` 输入 */
export const AstGrepFindByQueryInput = Schema.Struct({
  /** 工作区根目录 */
  workspaceRoot: TrimmedNonEmptyString,
  /** 目标语言 */
  language: AstGrepLanguage,
  /** S-expression 查询字符串 */
  query: TrimmedNonEmptyString,
});
export type AstGrepFindByQueryInput = typeof AstGrepFindByQueryInput.Type;

/** `indexer.astGrepFindByName` 输入（统一 find_references + find_calls_to 入口） */
export const AstGrepFindByNameInput = Schema.Struct({
  /** 工作区根目录 */
  workspaceRoot: TrimmedNonEmptyString,
  /** 名称（支持 `"foo"` / `"obj.foo"` 两种形式） */
  name: TrimmedNonEmptyString,
  /** `references`：找所有标识符出现位置；`calls`：仅找 `name(...)` 形式调用（默认 `calls`） */
  mode: Schema.optional(AstGrepFindMode),
});
export type AstGrepFindByNameInput = typeof AstGrepFindByNameInput.Type;

/** `indexer.astGrepListPresets` 输入 */
export const AstGrepListPresetsInput = Schema.Struct({
  /** 仅返回支持目标语言的 preset（None 表示全部） */
  language: Schema.optional(AstGrepLanguage),
});
export type AstGrepListPresetsInput = typeof AstGrepListPresetsInput.Type;

/** `indexer.astGrepCompilePattern` 输入 */
export const AstGrepCompilePatternInput = Schema.Struct({
  /** 目标语言 */
  language: AstGrepLanguage,
  /** 用户友好的模式（如 `"console.log($MSG)"` / `"call_expression"`） */
  pattern: TrimmedNonEmptyString,
});
export type AstGrepCompilePatternInput = typeof AstGrepCompilePatternInput.Type;

// ===========================================================================
// 结构化重写
// ===========================================================================

/** `indexer.astGrepRewrite` 输入 */
export const AstGrepRewriteInput = Schema.Struct({
  /** 目标文件路径（绝对路径或相对工作区根） */
  filePath: TrimmedNonEmptyString,
  /** 目标语言 */
  language: AstGrepLanguage,
  /** 匹配模式 */
  pattern: TrimmedNonEmptyString,
  /** 替换模板（支持 `$NAME` / `$$$BODY` 透传） */
  rewrite: TrimmedNonEmptyString,
  /** 是否仅预览不落盘（默认 false）
   *
   * - `true`：仅返回 `newContent`，不写盘，由调用方决定是否落盘
   * - `false` / `undefined`：替换后直接写回文件（旧行为）
   */
  dryRun: Schema.optional(Schema.Boolean),
});
export type AstGrepRewriteInput = typeof AstGrepRewriteInput.Type;

/** 单次替换位置 */
export const AstGrepRewriteLocation = Schema.Struct({
  file: TrimmedNonEmptyString,
  line: Schema.Number,
  column: Schema.Number,
});
export type AstGrepRewriteLocation = typeof AstGrepRewriteLocation.Type;

/** 替换结果 */
export const AstGrepRewriteResult = Schema.Struct({
  /** 替换后的完整内容 */
  newContent: Schema.String,
  /** 实际替换的次数 */
  replacements: Schema.Number,
  /** 命中的位置（file:line:column）— 用于 UI 展示 */
  matchLocations: Schema.Array(AstGrepRewriteLocation),
});
export type AstGrepRewriteResult = typeof AstGrepRewriteResult.Type;

File: ydsz-shared/src/ast_grep.rs
//! # AST-Grep 深化集成（P1-6）
//!
//! 结构化代码搜索和批量替换的后端基础设施。
//!
//! ## 核心能力
//!
//! - 按节点类型搜索（find by node kind）
//! - S-expression 模式匹配
//! - 按名称查找调用/引用
//! - 结构化重写（rewrite with pattern）
//! - 预设模式管理

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ============================================================================
// 预设模式
// ============================================================================

/// AST-Grep 预设模式 ID
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AstGrepPreset {
    ConsoleLog,
    ConsoleError,
    AwaitFetch,
    TryCatch,
    TodoComment,
    RustUnwrap,
    RustExpect,
    PyPrint,
    PyExcept,
}

impl AstGrepPreset {
    /// 获取所有预设
    pub fn all() -> &'static [AstGrepPreset] {
        &[
            AstGrepPreset::ConsoleLog,
            AstGrepPreset::ConsoleError,
            AstGrepPreset::AwaitFetch,
            AstGrepPreset::TryCatch,
            AstGrepPreset::TodoComment,
            AstGrepPreset::RustUnwrap,
            AstGrepPreset::RustExpect,
            AstGrepPreset::PyPrint,
            AstGrepPreset::PyExcept,
        ]
    }

    /// 获取预设的显示名称
    pub fn display_name(&self) -> &'static str {
        match self {
            AstGrepPreset::ConsoleLog => "console.log(...) 调用",
            AstGrepPreset::ConsoleError => "console.error(...) 调用",
            AstGrepPreset::AwaitFetch => "await fetch(...) 调用",
            AstGrepPreset::TryCatch => "try-catch 语句",
            AstGrepPreset::TodoComment => "TODO/FIXME 注释",
            AstGrepPreset::RustUnwrap => "Rust .unwrap() 调用",
            AstGrepPreset::RustExpect => "Rust .expect() 调用",
            AstGrepPreset::PyPrint => "Python print(...) 调用",
            AstGrepPreset::PyExcept => "Python except 异常处理",
        }
    }

    /// 获取预设支持的语言
    pub fn supported_languages(&self) -> &'static [&'static str] {
        match self {
            AstGrepPreset::ConsoleLog => &["typescript", "javascript"],
            AstGrepPreset::ConsoleError => &["typescript", "javascript"],
            AstGrepPreset::AwaitFetch => &["typescript", "javascript"],
            AstGrepPreset::TryCatch => &["typescript", "javascript", "rust", "python"],
            AstGrepPreset::TodoComment => &["typescript", "javascript", "rust", "python"],
            AstGrepPreset::RustUnwrap => &["rust"],
            AstGrepPreset::RustExpect => &["rust"],
            AstGrepPreset::PyPrint => &["python"],
            AstGrepPreset::PyExcept => &["python"],
        }
    }

    /// 获取预设对应的 S-expression 模式
    pub fn sexp_pattern(&self) -> &'static str {
        match self {
            AstGrepPreset::ConsoleLog => "console.log($$$ARGS)",
            AstGrepPreset::ConsoleError => "console.error($$$ARGS)",
            AstGrepPreset::AwaitFetch => "await fetch($URL)",
            AstGrepPreset::TryCatch => "try_statement",
            AstGrepPreset::TodoComment => "$COMMENT",
            AstGrepPreset::RustUnwrap => ".unwrap()",
            AstGrepPreset::RustExpect => ".expect($MSG)",
            AstGrepPreset::PyPrint => "print($$$ARGS)",
            AstGrepPreset::PyExcept => "except_clause",
        }
    }

    /// 检查是否支持指定语言
    pub fn supports_language(&self, lang: &str) -> bool {
        self.supported_languages().contains(&lang)
    }
}

impl std::fmt::Display for AstGrepPreset {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AstGrepPreset::ConsoleLog => write!(f, "console_log"),
            AstGrepPreset::ConsoleError => write!(f, "console_error"),
            AstGrepPreset::AwaitFetch => write!(f, "await_fetch"),
            AstGrepPreset::TryCatch => write!(f, "try_catch"),
            AstGrepPreset::TodoComment => write!(f, "todo_comment"),
            AstGrepPreset::RustUnwrap => write!(f, "rust_unwrap"),
            AstGrepPreset::RustExpect => write!(f, "rust_expect"),
            AstGrepPreset::PyPrint => write!(f, "py_print"),
            AstGrepPreset::PyExcept => write!(f, "py_except"),
        }
    }
}

/// 预设详情
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepPresetInfo {
    pub id: String,
    pub display_name: String,
    pub supported_languages: Vec<String>,
}

impl From<AstGrepPreset> for AstGrepPresetInfo {
    fn from(preset: AstGrepPreset) -> Self {
        Self {
            id: preset.to_string(),
            display_name: preset.display_name().to_string(),
            supported_languages: preset
                .supported_languages()
                .iter()
                .map(|s| s.to_string())
                .collect(),
        }
    }
}

// ============================================================================
// 目标语言
// ============================================================================

/// 目标语言
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AstGrepLanguage {
    Typescript,
    Javascript,
    Rust,
    Python,
}

impl std::fmt::Display for AstGrepLanguage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AstGrepLanguage::Typescript => write!(f, "typescript"),
            AstGrepLanguage::Javascript => write!(f, "javascript"),
            AstGrepLanguage::Rust => write!(f, "rust"),
            AstGrepLanguage::Python => write!(f, "python"),
        }
    }
}

// ============================================================================
// 匹配结果
// ============================================================================

/// 捕获项
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepCapture {
    pub name: String,
    pub text: String,
}

/// 单条匹配命中
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepMatch {
    /// 源文件路径
    pub file: String,
    /// 起始行（1-based）
    pub line: u32,
    /// 起始列（1-based）
    pub column: u32,
    /// 起始字节偏移
    pub start_byte: u32,
    /// 结束字节偏移
    pub end_byte: u32,
    /// 匹配的原始文本
    pub text: String,
    /// 节点类型
    pub node_kind: String,
    /// 捕获项
    #[serde(default)]
    pub captures: Vec<AstGrepCapture>,
}

/// 搜索结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepSearchResult {
    /// 匹配条目
    pub matches: Vec<AstGrepMatch>,
    /// 总匹配数
    pub total: usize,
    /// 搜索耗时（毫秒）
    pub elapsed_ms: u64,
}

impl AstGrepSearchResult {
    /// 创建空结果
    pub fn empty() -> Self {
        Self {
            matches: Vec::new(),
            total: 0,
            elapsed_ms: 0,
        }
    }

    /// 获取匹配的文件列表（去重）
    pub fn files(&self) -> Vec<&str> {
        let mut files: Vec<&str> = self
            .matches
            .iter()
            .map(|m| m.file.as_str())
            .collect();
        files.sort_unstable();
        files.dedup();
        files
    }

    /// 按文件分组
    pub fn group_by_file(&self) -> HashMap<&str, Vec<&AstGrepMatch>> {
        let mut groups: HashMap<&str, Vec<&AstGrepMatch>> = HashMap::new();
        for m in &self.matches {
            groups.entry(&m.file).or_default().push(m);
        }
        groups
    }
}

// ============================================================================
// 模式编译
// ============================================================================

/// 编译后的模式
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepCompiledPattern {
    /// S-expression 查询字符串
    pub query: String,
    /// 捕获名列表
    pub captures: Vec<String>,
}

/// 模式编译器
///
/// 将用户友好的模式转换为 tree-sitter S-expression。
pub struct AstGrepPatternCompiler;

impl AstGrepPatternCompiler {
    /// 编译模式
    pub fn compile(_language: AstGrepLanguage, pattern: &str) -> Result<AstGrepCompiledPattern, String> {
        // 如果已经是 S-expression 形式（包含括号），直接返回
        if pattern.starts_with('(') && pattern.ends_with(')') {
            let captures = Self::extract_captures(pattern);
            return Ok(AstGrepCompiledPattern {
                query: pattern.to_string(),
                captures,
            });
        }

        // 如果是简化的函数调用形式（如 "console.log($MSG)"），转换为 S-expression
        if let Some(sexp) = Self::convert_call_pattern(pattern) {
            let captures = Self::extract_captures(&sexp);
            return Ok(AstGrepCompiledPattern {
                query: sexp,
                captures,
            });
        }

        // 否则当作节点类型名直接匹配
        Ok(AstGrepCompiledPattern {
            query: pattern.to_string(),
            captures: Vec::new(),
        })
    }

    /// 从 S-expression 中提取捕获名
    fn extract_captures(sexp: &str) -> Vec<String> {
        let mut captures = Vec::new();
        for word in sexp.split(|c: char| !c.is_alphanumeric() && c != '_' && c != '$') {
            if word.starts_with('@') || word.starts_with('$') {
                captures.push(word.to_string());
            }
        }
        captures.sort_unstable();
        captures.dedup();
        captures
    }

    /// 转换简化调用模式为 S-expression
    fn convert_call_pattern(pattern: &str) -> Option<String> {
        // 简化转换：识别 "obj.method(ARGS)" 形式
        if pattern.contains('(') && pattern.contains(')') {
            let pattern = pattern.trim();
            return Some(pattern.to_string());
        }
        None
    }
}

// ============================================================================
// 结构化重写
// ============================================================================

/// 重写位置
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepRewriteLocation {
    pub file: String,
    pub line: u32,
    pub column: u32,
}

/// 重写结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepRewriteResult {
    /// 替换后的内容
    pub new_content: String,
    /// 替换次数
    pub replacements: u32,
    /// 命中位置
    pub match_locations: Vec<AstGrepRewriteLocation>,
}

/// 批量重写请求
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepBatchRewriteRequest {
    /// 工作区根目录
    pub workspace_root: String,
    /// 目标语言
    pub language: AstGrepLanguage,
    /// 匹配模式
    pub pattern: String,
    /// 替换模板
    pub rewrite: String,
    /// 文件过滤（Glob 模式）
    #[serde(default)]
    pub file_filter: Option<String>,
    /// 是否仅预览
    #[serde(default)]
    pub dry_run: bool,
}

/// 批量重写结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AstGrepBatchRewriteResult {
    /// 每个文件的结果
    pub file_results: Vec<AstGrepRewriteResult>,
    /// 总替换次数
    pub total_replacements: u32,
    /// 处理的文件数
    pub files_processed: u32,
}

/// AST-Grep 搜索引擎
///
/// 提供结构化的代码搜索和替换能力。
/// 实际实现依赖 sg (ast-grep) CLI 或库的集成。
pub struct AstGrepEngine;

impl AstGrepEngine {
    /// 按节点类型搜索
    pub fn find_by_node_kind(_workspace_root: &str, _kind: &str) -> AstGrepSearchResult {
        // 实际实现应调用 ast-grep CLI
        AstGrepSearchResult::empty()
    }

    /// 按 S-expression 模式搜索
    pub fn find_by_query(
        _workspace_root: &str,
        _language: AstGrepLanguage,
        _query: &str,
    ) -> AstGrepSearchResult {
        AstGrepSearchResult::empty()
    }

    /// 按名称查找
    pub fn find_by_name(
        _workspace_root: &str,
        _name: &str,
        _mode: &str,
    ) -> AstGrepSearchResult {
        AstGrepSearchResult::empty()
    }

    /// 结构化重写
    pub fn rewrite(
        _file_path: &str,
        _language: AstGrepLanguage,
        _pattern: &str,
        _rewrite: &str,
        _dry_run: bool,
    ) -> Result<AstGrepRewriteResult, String> {
        Err("AST-Grep 引擎需要安装 ast-grep CLI 工具".to_string())
    }

    /// 批量重写
    pub fn batch_rewrite(
        _request: &AstGrepBatchRewriteRequest,
    ) -> Result<AstGrepBatchRewriteResult, String> {
        Err("AST-Grep 引擎需要安装 ast-grep CLI 工具".to_string())
    }

    /// 获取推荐预设
    pub fn recommended_presets(language: AstGrepLanguage) -> Vec<AstGrepPresetInfo> {
        AstGrepPreset::all()
            .iter()
            .filter(|p| p.supports_language(&language.to_string()))
            .cloned()
            .map(AstGrepPresetInfo::from)
            .collect()
    }
}

// ============================================================================
// 测试
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_preset_display_names() {
        assert_eq!(AstGrepPreset::ConsoleLog.display_name(), "console.log(...) 调用");
        assert_eq!(AstGrepPreset::RustUnwrap.display_name(), "Rust .unwrap() 调用");
    }

    #[test]
    fn test_preset_language_support() {
        assert!(AstGrepPreset::ConsoleLog.supports_language("typescript"));
        assert!(AstGrepPreset::ConsoleLog.supports_language("javascript"));
        assert!(!AstGrepPreset::ConsoleLog.supports_language("rust"));

        assert!(AstGrepPreset::RustUnwrap.supports_language("rust"));
        assert!(!AstGrepPreset::RustUnwrap.supports_language("python"));
    }

    #[test]
    fn test_preset_sexp_patterns() {
        assert_eq!(AstGrepPreset::ConsoleLog.sexp_pattern(), "console.log($$$ARGS)");
        assert_eq!(AstGrepPreset::AwaitFetch.sexp_pattern(), "await fetch($URL)");
    }

    #[test]
    fn test_pattern_compiler_sexp() {
        let compiled = AstGrepPatternCompiler::compile(
            AstGrepLanguage::Typescript,
            "console.log($MSG)",
        ).unwrap();

        assert_eq!(compiled.query, "console.log($MSG)");
        assert!(compiled.captures.contains(&"$MSG".to_string()));
    }

    #[test]
    fn test_pattern_compiler_node_kind() {
        let compiled = AstGrepPatternCompiler::compile(
            AstGrepLanguage::Typescript,
            "try_statement",
        ).unwrap();

        assert_eq!(compiled.query, "try_statement");
        assert!(compiled.captures.is_empty());
    }

    #[test]
    fn test_search_result_group_by_file() {
        let result = AstGrepSearchResult {
            matches: vec![
                AstGrepMatch {
                    file: "src/a.ts".to_string(),
                    line: 10,
                    column: 5,
                    start_byte: 100,
                    end_byte: 120,
                    text: "console.log('a')".to_string(),
                    node_kind: "call_expression".to_string(),
                    captures: vec![],
                },
                AstGrepMatch {
                    file: "src/a.ts".to_string(),
                    line: 20,
                    column: 3,
                    start_byte: 200,
                    end_byte: 220,
                    text: "console.log('b')".to_string(),
                    node_kind: "call_expression".to_string(),
                    captures: vec![],
                },
                AstGrepMatch {
                    file: "src/b.ts".to_string(),
                    line: 5,
                    column: 1,
                    start_byte: 50,
                    end_byte: 70,
                    text: "console.log('c')".to_string(),
                    node_kind: "call_expression".to_string(),
                    captures: vec![],
                },
            ],
            total: 3,
            elapsed_ms: 15,
        };

        let groups = result.group_by_file();
        assert_eq!(groups.len(), 2);
        assert_eq!(groups.get("src/a.ts").unwrap().len(), 2);
        assert_eq!(groups.get("src/b.ts").unwrap().len(), 1);
    }

    #[test]
    fn test_recommended_presets() {
        let ts_presets = AstGrepEngine::recommended_presets(AstGrepLanguage::Typescript);
        assert!(ts_presets.iter().any(|p| p.id == "console_log"));
        assert!(ts_presets.iter().any(|p| p.id == "try_catch"));
        assert!(!ts_presets.iter().any(|p| p.id == "rust_unwrap"));

        let rust_presets = AstGrepEngine::recommended_presets(AstGrepLanguage::Rust);
        assert!(rust_presets.iter().any(|p| p.id == "rust_unwrap"));
        assert!(rust_presets.iter().any(|p| p.id == "try_catch"));
        assert!(!rust_presets.iter().any(|p| p.id == "console_log"));
    }

    #[test]
    fn test_preset_from_string() {
        // 验证预设名称序列化
        assert_eq!(AstGrepPreset::ConsoleLog.to_string(), "console_log");
        assert_eq!(AstGrepPreset::RustUnwrap.to_string(), "rust_unwrap");
        assert_eq!(AstGrepPreset::PyPrint.to_string(), "py_print");
    }
}
